content/articles/mocking-with-junit-and-mockito-the-why-and-how/index.md

```diff
@@ -0,0 +1,1011 @@
+  assertThat(studentDto, hasProperty("lastName", equalTo("Doe")));
+  assertThat(studentDto, hasProperty("matricNo", equalTo("MAT100419")));
+  assertThat(studentDto, hasProperty("gender", equalTo(Gender.MALE)));
+}
+```
+Next, consider the same test but using an ```ArgumentCaptor``` instead:
+
+```java
+@Test
+void registerStudentTest() throws Exception {
+//these parts are skipped for brevity sake
+  when(studentRepository.save(studentArgumentCaptor.capture())).thenReturn(studentToSave);
+  StudentDto studentDto = studentService.registerStudent(registrationRequest);
+  assertThat(studentDto, hasProperty("firstName", equalTo("John")));
+  assertThat(studentDto, hasProperty("lastName", equalTo("Doe")));
+  assertThat(studentDto, hasProperty("matricNo", equalTo("MAT100419")));
+  assertThat(studentDto, hasProperty("gender", equalTo(Gender.MALE)));
+}
+```
+
+As demonstrated above, the readability of our tests is reduced when we use argument captors instead of argument matchers in stubbing methods.
+
+Also, if the ```studentService.registerStudent(registrationRequest)``` does not call ```studentRepository.save(student)```, we will get an exception:
+
+>org.mockito.exceptions.base.MockitoException:
+No argument value was captured!
+
+We get this error because the stubbed method has not captured an argument. We used ```ArgumentCaptor``` instead of an Argument Matcher when stubbing methods misdirects to an exception in the test instead of the method we are testing.
+
+
+### Conclusion
+In this tutorial, we have successfully learned how to write unit tests using Mockito, the different types of mocking frameworks, the different ways of creating mock objects using Mockito, and how to perform basic stubbing in Mockito using the ```when- then``` directive. Finally, we concluded with learning how to use argument captors.
+
+The code is accessible from this [repository](https://github.com/ehizman/bedspaces)
+
+Happy Coding!
+
+### References
+1. Using Mockito ArgumentCaptor- https://www.baeldung.com/mockito-argumentcaptor
+2. Testing with Hamcrest - https://www.baeldung.com/java-junit-hamcrest-guide
+3. Mockito ArgumentMatchers- https://www.baeldung.com/mockito-argument-matchers
+4. Mockito Verify Cookbook-https://www.baeldung.com/mockito-verify
+5. Stubbing and Mocking with Mockito and JUnit-https://semaphoreci.com/community/tutorials/stubbing-and-mocking-with-mockito-2-and-junit
+6. Interface InOrder-https://www.javadoc.io/doc/org.mockito/mockito-core/2.6.9/org/mockito/InOrder.html
+7. Mockito - verifyNoMoreInteractions() and verifyNoInteractions()-https://www.logicbig.com/tutorials/unit-testing/mockito/verify-no-more-interactions.html
+8. Mastering Unit Testing Using Mockito and JUnit -An advanced guide to mastering unit testing using Mockito and JUnit by Sujoy Acharya
+9. Interface VerificationMode - https://www.javadoc.io/doc/org.mockito/mockito-core/2.2.6/org/mockito/verification/VerificationMode.html
```

Comment:
You can use this format to include your references:

![link](https://user-images.githubusercontent.com/50293753/147734281-83f6f102-8a1d-46d7-8d29-a8733aaa0ff1.png)

 So that it can look like this:

1. [Using Mockito ArgumentCaptor](https://www.baeldung.com/mockito-argumentcaptor)